transmogrifier/sources/marc.py

```diff
@@ -0,0 +1,200 @@
+
+        # contributors
+
+        # dates
+
+        # edition
+
+        # file_formats
+
+        # format
+
+        # funding_information
+
+        # holdings
+
+        # identifiers
+
+        # languages
+
+        # links
+
+        # literary_form
+        # Literary form is applicable to Book (BK) material configurations and indicated
+        # by leader "Type of Record" position = "Language Material" or "Manuscript
+        # language material" and "Bibliographic level" position =
+        # "Monographic component part," "Collection," "Subunit," or "Monograph/Item."
+        if leader := xml.find("leader", string=True):
+            if leader.string[6:7] in "at" and leader.string[7:8] in "acdm":
+                if fixed_length_data := xml.find(
+                    "controlfield", tag="008", string=True
+                ):
+                    if fixed_length_data.string[33:34] in "0se":
+                        fields["literary_form"] = "Nonfiction"
+                    elif fixed_length_data.string[33:34]:
+                        fields["literary_form"] = "Fiction"
+
+        # locations
+
+        # notes
+
+        # numbering
+
+        # physical_description
+
+        # publication_frequency
+
+        # publication_information
+
+        # related_items
+
+        # rights
+
+        # subjects
+
+        # summary
+
+        return fields
+
+    @staticmethod
+    def get_main_titles(xml: Tag) -> list[str]:
+        """
+        Retrieve main title(s) from a MARC XML record.
+
+        Overrides metaclass get_main_titles() method.
+
+        Args:
+            xml: A BeautifulSoup Tag representing a single MARC XML record.
+        """
+        main_title_values = [
+            title
+            for datafield in xml.find_all("datafield", tag="245")
+            if (
+                title := Marc.create_subfield_value_string_from_datafield(
+                    datafield, "abfgknps", " "
+                )
+            )
+        ]
+        if main_title_values:
+            return main_title_values
+        else:
+            return []
+
+    @staticmethod
+    def get_source_record_id(xml: Tag) -> str:
+        """
+        Get the source record ID from a MARC XML record.
+
+        Overrides metaclass get_source_record_id() method.
+
+        Args:
+            xml: A BeautifulSoup Tag representing a single MARC XML record.
+        """
+        return xml.find("controlfield", tag="001", string=True).string
+
+    @staticmethod
+    def create_subfield_value_list_from_datafield(
+        xml_element: Tag,
+        subfield_codes: str,
+    ) -> list:
+        """
+        Create a list of values from the specified subfields of a
+        datafield element.
+
+        Args:
+            xml_element: A BeautifulSoup Tag representing a single MARC XML element.
+            subfield_codes: The codes of the subfields to extract.
+        """
+        value_list = []
+        for subfield in xml_element.children:
```

Comment:
I think you could simplify this by just searching for what you want:
```
for subfield in xml_element.find_all(True, string=True):
    etc.
```
Then you only have to check for the right code.